python3Packages.parse: 1.16.0 -> 1.18.0, python-docx: run behave tests

maxxk opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

Fix `behave` which was broken from previous `parse` update (see `nixpkgs-review` from #95690; https://github.com/behave/behave/issues/862).

ZHF: #97479

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

nixpkgs-review:

```
3 packages failed to build:
python37Packages.pytest-bdd python38Packages.pytest-bdd terraform-compliance

17 packages built:
nrfutil python27Packages.parse python27Packages.parse-type python27Packages.pyparser python27Packages.pytest-bdd python37Packages.behave python37Packages.parse python37Packages.parse-type python37Packages.pyparser python37Packages.python-docx python37Packages.radish-bdd python38Packages.behave python38Packages.parse python38Packages.parse-type python38Packages.pyparser python38Packages.python-docx python38Packages.radish-bdd
```

Failed packages also fail on master.

maxxk committed:
python3Packages.parse: 1.16.0 -> 1.18.0

maxxk committed:
python3Packages.python-docx: run behave tests

Previously the problematic "behave" dependency
was not even used at checkPhase.

risicle commented:
`behave` fails to build for me on macos 10.14:

```
=================================== FAILURES ===================================
___ TestAsyncStepDecoratorPy34.test_step_decorator_async_run_until_complete2 ___

self = <tests.api._test_async_step34.TestAsyncStepDecoratorPy34 object at 0x1053c5150>

    def test_step_decorator_async_run_until_complete2(self):
        step_container = SimpleStepContainer()
        with use_step_import_modules(step_container):
            # -- STEP-DEFINITIONS EXAMPLE (as MODULE SNIPPET):
            # VARIANT 2: Use @asyncio.coroutine def step_impl()
            from behave import step
            from behave.api.async_step import async_run_until_complete
            import asyncio
    
            @step('a tagged-coroutine async step waits "{duration:f}" seconds')
            @async_run_until_complete
            @asyncio.coroutine
            def step_async_step_waits_seconds2(context, duration):
                yield from asyncio.sleep(duration)
    
        # -- USES: async def step_impl(...) as async-step (coroutine)
        AsyncStepTheory.validate(step_async_step_waits_seconds2)
    
        # -- RUN ASYNC-STEP: Verify that it is behaving correctly.
        # ENSURE: Execution of async-step matches expected duration.
        context = Context(runner=Runner(config={}))
        with StopWatch() as stop_watch:
            step_async_step_waits_seconds2(context, duration=0.2)
>       assert abs(stop_watch.duration - 0.2) <= 0.05
E       AssertionError

tests/api/_test_async_step34.py:75: AssertionError
___ TestAsyncStepDecoratorPy35.test_step_decorator_async_run_until_complete1 ___

self = <tests.api._test_async_step35.TestAsyncStepDecoratorPy35 object at 0x1051bd890>

    def test_step_decorator_async_run_until_complete1(self):
        step_container = SimpleStepContainer()
        with use_step_import_modules(step_container):
            # -- STEP-DEFINITIONS EXAMPLE (as MODULE SNIPPET):
            # VARIANT 1: Use async def step_impl()
            from behave import step
            from behave.api.async_step import async_run_until_complete
            import asyncio
    
            @step('an async coroutine step waits "{duration:f}" seconds')
            @async_run_until_complete
            async def step_async_step_waits_seconds(context, duration):
                await asyncio.sleep(duration)
    
        # -- USES: async def step_impl(...) as async-step (coroutine)
        AsyncStepTheory.validate(step_async_step_waits_seconds)
    
        # -- RUN ASYNC-STEP: Verify that it is behaving correctly.
        # ENSURE: Execution of async-step matches expected duration.
        context = Context(runner=Runner(config={}))
        with StopWatch() as stop_watch:
            step_async_step_waits_seconds(context, 0.2)
>       assert abs(stop_watch.duration - 0.2) <= 0.05
E       AssertionError

tests/api/_test_async_step35.py:75: AssertionError
=============================== warnings summary ===============================
tests/unit/test_behave4cmd_command_shell_proc.py:26
  /private/tmp/nix-build-behave-1.2.7.dev1.drv-1/source/tests/unit/test_behave4cmd_command_shell_proc.py:26: DeprecationWarning: invalid escape sequence \w
    winpath_pattern = u"^([A-Za-z]:(\\[\w\.\-]+)+)|((\\[\w\.\-]+)*)|(\s[\w\.\-]+([\w\.\-]+)*)$"

-- Docs: https://docs.pytest.org/en/latest/warnings.html
- generated xml file: /private/tmp/nix-build-behave-1.2.7.dev1.drv-1/source/build/testing/report.xml -
- generated html file: file:///private/tmp/nix-build-behave-1.2.7.dev1.drv-1/source/build/testing/report.html -
=========================== short test summary info ============================
FAILED tests/api/test_async_step.py::TestAsyncStepDecoratorPy34::test_step_decorator_async_run_until_complete2
FAILED tests/api/test_async_step.py::TestAsyncStepDecoratorPy35::test_step_decorator_async_run_until_complete1
======= 2 failed, 1248 passed, 7 skipped, 2 xfailed, 1 warning in 12.80s =======
```

This looks like its a timing-based test, which are always trouble - perhaps it should be disabled?

drewrisinger reviewed (approved):
* [x] Diff LGTM
* [x] Commits LGTM
* [x] Confirm nix-review results from author:

```
https://github.com/NixOS/nixpkgs/pull/97737
2 packages failed to build:
python38Packages.pytest-bdd terraform-compliance

18 packages built:
nrfutil python27Packages.parse python27Packages.parse-type python27Packages.pyparser python27Packages.pytest-bdd python37Packages.behave python37Packages.parse python37Packages.parse-type python37Packages.pyparser python37Packages.pytest-bdd python37Packages.python-docx python37Packages.radish-bdd python38Packages.behave python38Packages.parse python38Packages.parse-type python38Packages.pyparser python38Packages.python-docx python38Packages.radish-bdd
```

maxxk commented:
> `behave` fails to build for me on macos 10.14:

@risicle I tried several times building both `python38Packages.behave` and, as in your example, `python37Packages.behave` but unfortunately I couldn't reproduce the issue using fresh Nix install on macOS 10.14.6 18G6020 :( 

Probably we could ping someone maintaining nixpkgs on darwin to continue investigating?

In the meantime, could you please check that changing `pytest` command to the following fixes the build for you:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/behave/default.nix#L33
```
pytest tests -k "not test_step_decorator_async_run_until_complete"
```

risicle commented:
I simply think I've got slower computers than you (maybe try building on a heavily loaded machine), and as I say, timing-based tests are just trouble sooner or later. It's not too much of a loss disabling just `test_step_decorator_async_run_until_complete*` is it?

drewrisinger commented:
You can disable only that test if using ``pytestCheckHook`` and ``disableTests``

```nix
checkInputs = [ pytestCheckHook ... ];

disabledTests = lib.optionals stdenv.isDarwin [ "test_step_decorator_async_run_until_complete" ];
```

maxxk commented:
Thanks, I added the change. I'd just prefer to fix the test, at least by increasing the timeout, but skipping it on darwin is fine too.

drewrisinger commented on pkgs/development/python-modules/behave/default.nix:
```diff
@@ -16,22 +16,21 @@ buildPythonApplication rec {
   '';
 
-  doCheck = true;
+  pytestFlagsArray = [ "tests/" ];
```
Probably redundant b/c pytest would automatically find that folder, but not important.

jonringer commented on pkgs/development/python-modules/behave/default.nix:
```diff
@@ -16,22 +16,20 @@ buildPythonApplication rec {
   '';
 
-  doCheck = true;
+  disabledTests = stdenv.lib.optionals stdenv.isDarwin [ "test_step_decorator_async_run_until_complete" ];
```
do you mind adding a comment as to why disabling this test is necessary?

drewrisinger commented on pkgs/development/python-modules/behave/default.nix:
```diff
@@ -16,22 +16,22 @@ buildPythonApplication rec {
   };
 
-  checkInputs = [ mock pathpy pyhamcrest pytest pytest-html ];
+  checkInputs = [ mock pathpy pyhamcrest pytest pytest-html pytestCheckHook ];
```
```suggestion
  checkInputs = [ pytestCheckHook mock pathpy pyhamcrest pytest-html ];
```
``pytest`` is redundant with ``pytestCheckHook``. Personal preference is to put ``pytestCheckHook`` first b/c it's the test runner.

drewrisinger commented on pkgs/development/python-modules/behave/default.nix:
```diff
@@ -1,6 +1,6 @@
 { stdenv, fetchFromGitHub
 , buildPythonApplication, python
-, mock, pathpy, pyhamcrest, pytest, pytest-html
+, mock, pathpy, pyhamcrest, pytest, pytest-html, pytestCheckHook
```
```suggestion
, pytestCheckHook, mock, pathpy, pyhamcrest, pytest-html,
```

drewrisinger reviewed (changes requested):
* [ ] Minor point in diff still: redundant ``pytest``
* [x] Commits LGTM
* [x] Builds via ``nix-review``: Failed packages fail on ``master``

```
https://github.com/NixOS/nixpkgs/pull/97737
2 packages failed to build:
python38Packages.pytest-bdd terraform-compliance

18 packages built:
nrfutil python27Packages.parse python27Packages.parse-type python27Packages.pyparser python27Packages.pytest-bdd python37Packages.behave python37Packages.parse python37Packages.parse-type python37Packages.pyparser python37Packages.pytest-bdd python37Packages.python-docx python37Packages.radish-bdd python38Packages.behave python38Packages.parse python38Packages.parse-type python38Packages.pyparser python38Packages.python-docx python38Packages.radish-bdd
```

maxxk committed:
python3Packages.behave: skip test failing on darwin

Timing-based test is flaky on Darwin:
https://github.com/NixOS/nixpkgs/pull/97737#issuecomment-691489824

jonringer commented:
sorry @maxxk that this turned out to be such a ..... pain of a PR

but thanks for sticking through it :)

jonringer reviewed (approved):
LGTM

failures are broken on target branch
```
https://github.com/NixOS/nixpkgs/pull/97737
3 packages failed to build:
python37Packages.pytest-bdd python38Packages.pytest-bdd terraform-compliance

17 packages built:
nrfutil python27Packages.parse python27Packages.parse-type python27Packages.pyparser python27Packages.pytest-bdd python37Packages.behave python37Packages.parse python37Packages.parse-type python37Packages.pyparser python37Packages.python-docx python37Packages.radish-bdd python38Packages.behave python38Packages.parse python38Packages.parse-type python38Packages.pyparser python38Packages.python-docx python38Packages.radish-bdd
```
